File: src/Components/src/Composite/ValidationPackageSelector/Helper.test.ts
import { describe, it, expect } from 'vitest';
import {
  hasFlag,
  toggleFlag,
  toVersionString,
  rowState,
  filterBySearch,
  filterByTag,
  filterByAuthor,
  sortByChecked,
  pageCount,
  slicePage,
  unlistedNames,
  computeNewPackages,
} from './Helper.fs.ts';
import {
  ValidationPackageDTO_Create_Z1A22E2B7 as createDto,
  CheckedSort_None,
  CheckedSort_CheckedFirst,
  CheckedSort_CheckedLast,
  PackageRowState_Unchecked,
  PackageRowState_Checked,
  PackageRowState_HasOlderVersion,
} from './Types.fs.ts';
import type { ValidationPackageDTO, PackageRowState_$union } from './Types.fs.ts';
import { ValidationPackage } from '../../fable_modules/ARCtrl.ValidationPackages.3.0.0-beta.12/ValidationPackage.fs.js';
import { ValidationPackagesConfig } from '../../fable_modules/ARCtrl.ValidationPackages.3.0.0-beta.12/ValidationPackagesConfig.fs.js';
import { ofArray as mapOfArray } from '../../fable_modules/fable-library-ts.5.0.0-alpha.21/Map.ts';
import { ofArray as setOfArray } from '../../fable_modules/fable-library-ts.5.0.0-alpha.21/Set.ts';

const stringComparer = { Compare: (a: string, b: string) => a.localeCompare(b) };
const editsOf = (entries: Array<[string, ValidationPackage | null]>) =>
  mapOfArray<string, ValidationPackage | null>(entries, stringComparer);
const setOf = (values: string[]) => setOfArray(values, stringComparer);

const NAME = 1;
const SUMMARY = 2;
const DESCRIPTION = 4;
const TAGS = 8;
const AUTHORS = 16;

type Tag = { Name: string | null; TermSourceREF: string | null; TermAccessionNumber: string | null };
type Author = { FullName: string | null; Email: string | null; Affiliation: string | null; AffiliationLink: string | null };

const tag = (name: string): Tag => ({ Name: name, TermSourceREF: null, TermAccessionNumber: null });
const author = (fullName: string): Author => ({ FullName: fullName, Email: null, Affiliation: null, AffiliationLink: null });

function mkDto(overrides: Partial<Record<string, unknown>> & { Name?: string }): ValidationPackageDTO {
  return createDto(
    (overrides.Name as string) ?? 'Pkg',
    (overrides.Summary as string) ?? 'Summary text',
    (overrides.Description as string) ?? 'Description text',
    (overrides.MajorVersion as number) ?? 1,
    (overrides.MinorVersion as number) ?? 0,
    (overrides.PatchVersion as number) ?? 0,
    (overrides.PreReleaseVersionSuffix as string) ?? '',
    (overrides.BuildMetadataVersionSuffix as string) ?? '',
    [],
    new Date(),
    (overrides.Tags as Tag[]) ?? [],
    '',
    '',
    (overrides.Authors as Author[]) ?? [],
    'python'
  );
}

function mkConfig(packages: Array<[string, string | null]>): ValidationPackagesConfig {
  const arr = packages.map(([name, version]) => new ValidationPackage(name, version ?? null));
  return ValidationPackagesConfig.make(arr as any, null as any);
}

describe('hasFlag / toggleFlag', () => {
  it('hasFlag detects single and combined flags', () => {
    expect(hasFlag(NAME, NAME)).toBe(true);
    expect(hasFlag(NAME, SUMMARY)).toBe(false);
    expect(hasFlag(NAME + SUMMARY, SUMMARY)).toBe(true);
    expect(hasFlag(0, NAME)).toBe(false);
  });

  it('toggleFlag flips a flag', () => {
    expect(toggleFlag(NAME, SUMMARY)).toBe(NAME + SUMMARY);
    expect(toggleFlag(NAME + SUMMARY, SUMMARY)).toBe(NAME);
  });
});

describe('toVersionString', () => {
  it('formats plain versions', () => {
    expect(toVersionString(mkDto({ MajorVersion: 1, MinorVersion: 2, PatchVersion: 3 }))).toBe('1.2.3');
  });
  it('appends prerelease and build metadata', () => {
    expect(
      toVersionString(
        mkDto({ MajorVersion: 2, MinorVersion: 0, PatchVersion: 0, PreReleaseVersionSuffix: 'alpha.1', BuildMetadataVersionSuffix: '7' })
      )
    ).toBe('2.0.0-alpha.1+7');
  });
});

describe('rowState', () => {
  it('is Unchecked (0) when name absent', () => {
    expect(rowState(mkConfig([]), mkDto({ Name: 'A' })).tag).toBe(0);
  });
  it('is Checked (1) when version matches', () => {
    expect(rowState(mkConfig([['A', '1.0.0']]), mkDto({ Name: 'A' })).tag).toBe(1);
  });
  it('is HasOlderVersion (2) on mismatch or missing version', () => {
    expect(rowState(mkConfig([['A', '0.9.0']]), mkDto({ Name: 'A' })).tag).toBe(2);
    expect(rowState(mkConfig([['A', null]]), mkDto({ Name: 'A' })).tag).toBe(2);
  });
});

describe('filterBySearch', () => {
  const pkgs = [
    mkDto({ Name: 'Invenio', Summary: 'A great package', Tags: [tag('DataPLANT')] }),
    mkDto({ Name: 'Other', Summary: 'Invenio mentions', Authors: [author('Kevin Frey')] }),
    mkDto({ Name: 'Third', Summary: 'Nothing here' }),
  ];

  it('returns all for empty query', () => {
    expect(filterBySearch(NAME, '', pkgs)).toHaveLength(3);
  });
  it('returns all for zero flags', () => {
    expect(filterBySearch(0, 'Invenio', pkgs)).toHaveLength(3);
  });
  it('name-only search', () => {
    const result = filterBySearch(NAME, 'invenio', pkgs);
    expect(result.map((p) => p.Name)).toEqual(['Invenio']);
  });
  it('summary-only search', () => {
    const result = filterBySearch(SUMMARY, 'invenio mentions', pkgs);
    expect(result.map((p) => p.Name)).toEqual(['Other']);
  });
  it('combined name+tags search matches tag names', () => {
    const result = filterBySearch(NAME + TAGS, 'dataplant', pkgs);
    expect(result.map((p) => p.Name)).toEqual(['Invenio']);
  });
  it('author search matches full names', () => {
    const result = filterBySearch(AUTHORS, 'kevin', pkgs);
    expect(result.map((p) => p.Name)).toEqual(['Other']);
  });
});

describe('filterByTag / filterByAuthor', () => {
  const pkgs = [
    mkDto({ Name: 'A', Tags: [tag('X')], Authors: [author('Anna')] }),
    mkDto({ Name: 'B', Tags: [tag('Y')], Authors: [author('Ben')] }),
  ];

  it('no filter returns all', () => {
    expect(filterByTag(null as any, pkgs)).toHaveLength(2);
    expect(filterByAuthor(null as any, pkgs)).toHaveLength(2);
  });
  it('filters by exact tag and author', () => {
    expect(filterByTag('X' as any, pkgs).map((p) => p.Name)).toEqual(['A']);
    expect(filterByAuthor('Ben' as any, pkgs).map((p) => p.Name)).toEqual(['B']);
  });
});

describe('sortByChecked', () => {
  const stateOf = (name: string): PackageRowState_$union => {
    if (name === 'CheckedB') return PackageRowState_Checked();
    if (name === 'Old') return PackageRowState_HasOlderVersion();
    return PackageRowState_Unchecked();
  };
  const rowStateOf = (dto: ValidationPackageDTO) => stateOf(dto.Name);

  it('returns packages in original order for None', () => {
    const pkgs = [mkDto({ Name: 'U1' }), mkDto({ Name: 'CheckedB' }), mkDto({ Name: 'U2' })];
    expect(sortByChecked(CheckedSort_None(), rowStateOf, pkgs).map((p) => p.Name)).toEqual([
      'U1',
      'CheckedB',
      'U2',
    ]);
  });

  it('puts checked rows first for CheckedFirst', () => {
    const pkgs = [mkDto({ Name: 'U1' }), mkDto({ Name: 'CheckedB' }), mkDto({ Name: 'U2' })];
    expect(sortByChecked(CheckedSort_CheckedFirst(), rowStateOf, pkgs).map((p) => p.Name)).toEqual([
      'CheckedB',
      'U1',
      'U2',
    ]);
  });

  it('puts checked rows last for CheckedLast', () => {
    const pkgs = [mkDto({ Name: 'U1' }), mkDto({ Name: 'CheckedB' }), mkDto({ Name: 'U2' })];
    expect(sortByChecked(CheckedSort_CheckedLast(), rowStateOf, pkgs).map((p) => p.Name)).toEqual([
      'U1',
      'U2',
      'CheckedB',
    ]);
  });

  it('groups Checked before HasOlderVersion before Unchecked and keeps stable order', () => {
    const pkgs = [
      mkDto({ Name: 'U1' }),
      mkDto({ Name: 'Old' }),
      mkDto({ Name: 'CheckedB' }),
      mkDto({ Name: 'U2' }),
    ];
    expect(sortByChecked(CheckedSort_CheckedFirst(), rowStateOf, pkgs).map((p) => p.Name)).toEqual([
      'CheckedB',
      'Old',
      'U1',
      'U2',
    ]);
    expect(sortByChecked(CheckedSort_CheckedLast(), rowStateOf, pkgs).map((p) => p.Name)).toEqual([
      'U1',
      'U2',
      'Old',
      'CheckedB',
    ]);
  });
});

describe('pageCount / slicePage', () => {
  const pkgs = Array.from({ length: 45 }, (_, i) => mkDto({ Name: `P${i}` }));

  it('computes page counts', () => {
    expect(pageCount([] as ValidationPackageDTO[])).toBe(0);
    expect(pageCount(pkgs)).toBe(3);
  });
  it('slices pages', () => {
    expect(slicePage(pkgs, 0)).toHaveLength(20);
    expect(slicePage(pkgs, 2)).toHaveLength(5);
    expect(slicePage(pkgs, 3)).toHaveLength(0);
  });
});

describe('unlistedNames', () => {
  it('returns config names missing from table', () => {
    const config = mkConfig([
      ['A', '1.0.0'],
      ['Legacy', '1.0.0'],
    ]);
    const pkgs = [mkDto({ Name: 'A' })];
    expect(unlistedNames(config, pkgs)).toEqual(['Legacy']);
  });
});

describe('computeNewPackages', () => {
  const latest = (name: string, version: string) => new ValidationPackage(name, version);

  it('adds checked, removes unchecked, keeps unedited', () => {
    const config = mkConfig([
      ['Keep', '1.0.0'],
      ['RemoveMe', '1.0.0'],
    ]);
    const pkgs = [mkDto({ Name: 'Keep' }), mkDto({ Name: 'RemoveMe' }), mkDto({ Name: 'Fresh' })];
    const edits = editsOf([
      ['RemoveMe', null],
      ['Fresh', latest('Fresh', '1.0.0')],
    ]);
    const result = computeNewPackages(config, pkgs, edits as any, setOf([]));
    expect(result.map((p: any) => p.Name).sort()).toEqual(['Fresh', 'Keep']);
  });

  it('keeps old version unless consciously updated', () => {
    const config = mkConfig([['Old', '0.9.0']]);
    const pkgs = [mkDto({ Name: 'Old' })];
    const untouched = computeNewPackages(config, pkgs, editsOf([]) as any, setOf([]));
    expect(untouched[0].Version).toBe('0.9.0');

    const updated = computeNewPackages(config, pkgs, editsOf([['Old', latest('Old', '1.0.0')]]) as any, setOf([]));
    expect(updated[0].Version).toBe('1.0.0');
  });

  it('keeps unlisted, drops removed-unlisted', () => {
    const config = mkConfig([
      ['A', '1.0.0'],
      ['Ghost', '1.0.0'],
    ]);
    const pkgs = [mkDto({ Name: 'A' })];
    const kept = computeNewPackages(config, pkgs, editsOf([]) as any, setOf([]));
    expect(kept.map((p: any) => p.Name).sort()).toEqual(['A', 'Ghost']);

    const removed = computeNewPackages(config, pkgs, editsOf([]) as any, setOf(['Ghost']));
    expect(removed.map((p: any) => p.Name)).toEqual(['A']);
  });
});

File: src/Components/src/Composite/ValidationPackageSelector/Helper.test.fs
module internal Swate.Components.Tests.ValidationPackageSelector.Helper

open ARCtrl.ValidationPackages
open Swate.Components.Composite.ValidationPackageSelector.Helper
open Swate.Components.Composite.ValidationPackageSelector.Types
open Vitest
open ARCtrl.Helper.SemVer

let private mkTag (tagName: string) : OntologyAnnotationDTO = {
    Name = Some tagName
    TermSourceREF = None
    TermAccessionNumber = None
}

let private mkAuthor (fullName: string) : AuthorDTO = {
    FullName = Some fullName
    Email = None
    Affiliation = None
    AffiliationLink = None
}

type ValidationPackageDTO with

    static member mkDefault
        (
            ?name: string,
            ?summary: string,
            ?description,
            ?majorVersion: int,
            ?minorVersion: int,
            ?patchVersion: int,
            ?preReleaseVersionSuffix: string,
            ?buildMetadataVersionSuffix: string,
            ?tags: OntologyAnnotationDTO[],
            ?authors: AuthorDTO[]
        ) : ValidationPackageDTO =
        {
            Name = defaultArg name "Pkg"
            Summary = defaultArg summary "Summary text"
            Description = defaultArg description "Description text"
            MajorVersion = defaultArg majorVersion 1
            MinorVersion = defaultArg minorVersion 0
            PatchVersion = defaultArg patchVersion 0
            PreReleaseVersionSuffix = defaultArg preReleaseVersionSuffix ""
            BuildMetadataVersionSuffix = defaultArg buildMetadataVersionSuffix ""
            PackageContent = [||]
            ReleaseDate = System.DateTime.UtcNow
            Tags = defaultArg tags [||]
            ReleaseNotes = ""
            CQCHookEndpoint = ""
            Authors = defaultArg authors [||]
            ProgrammingLanguage = "python"
        }

let private mkConfig (packages: (string * string option)[]) =
    packages
    |> Array.map (fun (pkgName, version) -> ValidationPackage(pkgName, ?version = version))
    |> ResizeArray
    |> fun arr -> ValidationPackagesConfig.make arr None

let private latest (pkgName: string) (version: string) =
    ValidationPackage(pkgName, ?version = Some version)


Vitest.describe (
    "Ensure SemVer comparison works as expected",
    fun () ->
        Vitest.test (
            "SemVer comparison functions",
            fun () ->
                let v1 = ARCtrl.Helper.SemVer.SemVer.tryOfString "1.0.0"
                let v2 = ARCtrl.Helper.SemVer.SemVer.tryOfString "1.0.1"
                let v3 = ARCtrl.Helper.SemVer.SemVer.tryOfString "1.0.0-alpha"
                let v4 = ARCtrl.Helper.SemVer.SemVer.tryOfString "1.0.0+build"

                Vitest.expect(SemVer.isEqualWithoutBuild v1.Value v1.Value).toBe true
                Vitest.expect(SemVer.isEqualWithoutBuild v1.Value v4.Value).toBe true
                Vitest.expect(SemVer.isOlder v1.Value v2.Value).toBe true
                Vitest.expect(SemVer.isOlder v1.Value v1.Value).toBe false
                Vitest.expect(SemVer.isOlder v3.Value v1.Value).toBe true
                Vitest.expect(SemVer.isOlder v4.Value v1.Value).toBe false
        )
)

Vitest.describe (
    "hasFlag / toggleFlag",
    fun () ->

        Vitest.test (
            "hasFlag detects single and combined flags",
            fun () ->
                Vitest.expect(hasFlag SearchFields.Name SearchFields.Name).toBe true
                Vitest.expect(hasFlag SearchFields.Name SearchFields.Summary).toBe false
                Vitest.expect(hasFlag (SearchFields.Name ||| SearchFields.Summary) SearchFields.Summary).toBe true
                Vitest.expect(hasFlag (enum<SearchFields> 0) SearchFields.Name).toBe false
        )

        Vitest.test (
            "toggleFlag flips a flag",
            fun () ->
                Vitest
                    .expect(toggleFlag SearchFields.Name SearchFields.Summary)
                    .toEqual (SearchFields.Name ||| SearchFields.Summary)

                Vitest.expect(toggleFlag (SearchFields.Name ||| SearchFields.Summary) SearchFields.Summary).toEqual
                    SearchFields.Name
        )
)

Vitest.describe (
    "toVersionString",
    fun () ->

        Vitest.test (
            "formats plain versions",
            fun () ->
                Vitest
                    .expect(
                        toVersionString (
                            ValidationPackageDTO.mkDefault (majorVersion = 1, minorVersion = 2, patchVersion = 3)
                        )
                    )
                    .toBe ("1.2.3")
        )

        Vitest.test (
            "appends prerelease and build metadata",
            fun () ->
                Vitest
                    .expect(
                        toVersionString (
                            ValidationPackageDTO.mkDefault (
                                majorVersion = 2,
                                minorVersion = 0,
                                patchVersion = 0,
                                preReleaseVersionSuffix = "alpha.1",
                                buildMetadataVersionSuffix = "7"
                            )
                        )
                    )
                    .toBe ("2.0.0-alpha.1+7")
        )
)

Vitest.describe (
    "rowState",
    fun () ->

        Vitest.test (
            "is Unchecked when name absent",
            fun () ->
                Vitest.expect(rowState (mkConfig [||]) (ValidationPackageDTO.mkDefault (name = "A"))).toEqual
                    PackageRowState.Unchecked
        )

        Vitest.test (
            "is Checked when version matches",
            fun () ->
                Vitest
                    .expect(rowState (mkConfig [| "A", Some "1.0.0" |]) (ValidationPackageDTO.mkDefault (name = "A")))
                    .toEqual
                    PackageRowState.Checked
        )

        Vitest.test (
            "is HasOlderVersion on mismatch",
            fun () ->
                Vitest
                    .expect(rowState (mkConfig [| "A", Some "0.9.0" |]) (ValidationPackageDTO.mkDefault (name = "A")))
                    .toEqual
                    PackageRowState.HasOlderVersion
        )

        Vitest.test (
            "is Checked when config version is missing, as it assumes latest",

            fun () ->
                Vitest.expect(rowState (mkConfig [| "A", None |]) (ValidationPackageDTO.mkDefault (name = "A"))).toEqual
                    PackageRowState.Checked
        )

        Vitest.test (
            "is Invalid when config version is higher than package version",
            fun () ->
                Vitest
                    .expect(rowState (mkConfig [| "A", Some "99.0.0" |]) (ValidationPackageDTO.mkDefault (name = "A")))
                    .toEqual
                    PackageRowState.InvalidVersion
        )
)

Vitest.describe (
    "filterBySearch",
    fun () ->

        let pkgs = [|
            ValidationPackageDTO.mkDefault (
                name = "Invenio",
                summary = "A great package",
                tags = [| mkTag "DataPLANT" |]
            )
            ValidationPackageDTO.mkDefault (
                name = "Other",
                summary = "Invenio mentions",
                authors = [| mkAuthor "Kevin Frey" |]
            )
            ValidationPackageDTO.mkDefault (name = "Third", summary = "Nothing here")
        |]

        Vitest.test (
            "returns all for empty query",
            fun () -> Vitest.expect(filterBySearch SearchFields.Name "" pkgs).toHaveLength 3
        )

        Vitest.test (
            "returns all for zero flags",
            fun () -> Vitest.expect(filterBySearch (enum<SearchFields> 0) "Invenio" pkgs).toHaveLength 3
        )

        Vitest.test (
            "name-only search",
            fun () ->
                let result = filterBySearch SearchFields.Name "invenio" pkgs
                Vitest.expect(Array.map (fun p -> p.Name) result).toEqual [| "Invenio" |]
        )

        Vitest.test (
            "summary-only search",
            fun () ->
                let result = filterBySearch SearchFields.Summary "invenio mentions" pkgs
                Vitest.expect(Array.map (fun p -> p.Name) result).toEqual [| "Other" |]
        )

        Vitest.test (
            "combined name+tags search matches tag names",
            fun () ->
                let result =
                    filterBySearch (SearchFields.Name ||| SearchFields.Tags) "dataplant" pkgs

                Vitest.expect(Array.map (fun p -> p.Name) result).toEqual [| "Invenio" |]
        )

        Vitest.test (
            "author search matches full names",
            fun () ->
                let result = filterBySearch SearchFields.Authors "kevin" pkgs
                Vitest.expect(Array.map (fun p -> p.Name) result).toEqual [| "Other" |]
        )
)

Vitest.describe (
    "filterByTag / filterByAuthor",
    fun () ->

        let pkgs = [|
            ValidationPackageDTO.mkDefault (name = "A", tags = [| mkTag "X" |], authors = [| mkAuthor "Anna" |])
            ValidationPackageDTO.mkDefault (name = "B", tags = [| mkTag "Y" |], authors = [| mkAuthor "Ben" |])
        |]

        Vitest.test (
            "no filter returns all",
            fun () ->
                Vitest.expect(filterByTag None pkgs).toHaveLength 2
                Vitest.expect(filterByAuthor None pkgs).toHaveLength 2
        )

        Vitest.test (
            "filters by exact tag and author",
            fun () ->
                Vitest.expect(filterByTag (Some "X") pkgs |> Array.map (fun p -> p.Name)).toEqual [| "A" |]
                Vitest.expect(filterByAuthor (Some "Ben") pkgs |> Array.map (fun p -> p.Name)).toEqual [| "B" |]
        )
)

Vitest.describe (
    "sortByChecked",
    fun () ->

        let RowStateMap =
            Map.ofList [
                "CheckedB", PackageRowState.Checked
                "Old", PackageRowState.HasOlderVersion
            ]

        Vitest.test (
            "returns packages in original order for None",
            fun () ->
                let pkgs = [|
                    ValidationPackageDTO.mkDefault (name = "U1")
                    ValidationPackageDTO.mkDefault (name = "CheckedB")
                    ValidationPackageDTO.mkDefault (name = "U2")
                |]

                Vitest.expect(sortByChecked CheckedSort.None RowStateMap pkgs |> Array.map (fun p -> p.Name)).toEqual [|
                    "U1"
                    "CheckedB"
                    "U2"
                |]
        )

        Vitest.test (
            "puts checked rows first for CheckedFirst",
            fun () ->
                let pkgs = [|
                    ValidationPackageDTO.mkDefault (name = "U1")
                    ValidationPackageDTO.mkDefault (name = "CheckedB")
                    ValidationPackageDTO.mkDefault (name = "U2")
                |]

                Vitest
                    .expect(
                        sortByChecked CheckedSort.CheckedFirst RowStateMap pkgs
                        |> Array.map (fun p -> p.Name)
                    )
                    .toEqual
                    [| "CheckedB"; "U1"; "U2" |]
        )

        Vitest.test (
            "puts checked rows last for CheckedLast",
            fun () ->
                let pkgs = [|
                    ValidationPackageDTO.mkDefault (name = "U1")
                    ValidationPackageDTO.mkDefault (name = "CheckedB")
                    ValidationPackageDTO.mkDefault (name = "U2")
                |]

                Vitest
                    .expect(
                        sortByChecked CheckedSort.CheckedLast RowStateMap pkgs
                        |> Array.map (fun p -> p.Name)
                    )
                    .toEqual
                    [| "U1"; "U2"; "CheckedB" |]
        )

        Vitest.test (
            "groups Checked before HasOlderVersion before Unchecked and keeps stable order",
            fun () ->
                let pkgs = [|
                    ValidationPackageDTO.mkDefault (name = "U1")
                    ValidationPackageDTO.mkDefault (name = "Old")
                    ValidationPackageDTO.mkDefault (name = "CheckedB")
                    ValidationPackageDTO.mkDefault (name = "U2")
                |]

                Vitest
                    .expect(
                        sortByChecked CheckedSort.CheckedFirst RowStateMap pkgs
                        |> Array.map (fun p -> p.Name)
                    )
                    .toEqual
                    [| "CheckedB"; "Old"; "U1"; "U2" |]

                Vitest
                    .expect(
                        sortByChecked CheckedSort.CheckedLast RowStateMap pkgs
                        |> Array.map (fun p -> p.Name)
                    )
                    .toEqual
                    [| "U1"; "U2"; "Old"; "CheckedB" |]
        )
)

Vitest.describe (
    "pageCount / slicePage",
    fun () ->

        let pkgs = Array.init 45 (fun i -> ValidationPackageDTO.mkDefault (name = $"P{i}"))

        Vitest.test (
            "computes page counts",
            fun () ->
                Vitest.expect(pageCount [||]).toBe 0
                Vitest.expect(pageCount pkgs).toBe 3
        )

        Vitest.test (
            "slices pages",
            fun () ->
                Vitest.expect(slicePage pkgs 0).toHaveLength 20
                Vitest.expect(slicePage pkgs 2).toHaveLength 5
                Vitest.expect(slicePage pkgs 3).toHaveLength 0
        )
)

Vitest.describe (
    "unlistedNames",
    fun () ->

        Vitest.test (
            "returns config names missing from table",
            fun () ->
                let config = mkConfig [| "A", Some "1.0.0"; "Legacy", Some "1.0.0" |]
                let pkgs = [| ValidationPackageDTO.mkDefault (name = "A") |]
                Vitest.expect(unlistedNames config pkgs).toEqual [| "Legacy" |]
        )
)

Vitest.describe (
    "computeNewPackages",
    fun () ->

        Vitest.test (
            "adds checked, removes unchecked, keeps unedited",
            fun () ->
                let config = mkConfig [| "Keep", Some "1.0.0"; "RemoveMe", Some "1.0.0" |]

                let pkgs = [|
                    ValidationPackageDTO.mkDefault (name = "Keep")
                    ValidationPackageDTO.mkDefault (name = "RemoveMe")
                    ValidationPackageDTO.mkDefault (name = "Fresh")
                |]

                let edits = Map.ofList [ "RemoveMe", None; "Fresh", Some(latest "Fresh" "1.0.0") ]
                let result = computeNewPackages config pkgs edits Set.empty
                Vitest.expect(result |> Array.map (fun p -> p.Name) |> Array.sort).toEqual [| "Fresh"; "Keep" |]
        )

        Vitest.test (
            "keeps old version unless consciously updated",
            fun () ->
                let config = mkConfig [| "Old", Some "0.9.0" |]
                let pkgs = [| ValidationPackageDTO.mkDefault (name = "Old") |]

                let untouched = computeNewPackages config pkgs Map.empty Set.empty
                Vitest.expect(untouched.[0].Version).toBe (Some "0.9.0")

                let updated =
                    computeNewPackages config pkgs (Map.ofList [ "Old", Some(latest "Old" "1.0.0") ]) Set.empty

                Vitest.expect(updated.[0].Version).toBe (Some "1.0.0")
        )

        Vitest.test (
            "keeps unlisted, drops removed-unlisted",
            fun () ->
                let config = mkConfig [| "A", Some "1.0.0"; "Ghost", Some "1.0.0" |]
                let pkgs = [| ValidationPackageDTO.mkDefault (name = "A") |]

                let kept = computeNewPackages config pkgs Map.empty Set.empty
                Vitest.expect(kept |> Array.map (fun p -> p.Name) |> Array.sort).toEqual [| "A"; "Ghost" |]

                let removed = computeNewPackages config pkgs Map.empty (Set.ofList [ "Ghost" ])
                Vitest.expect(removed |> Array.map (fun p -> p.Name)).toEqual [| "A" |]
        )
)
